src/world-company-remuneration/salary-calculator-service-impl.ts

```diff
@@ -3,6 +3,20 @@ const K = "K";
  * @author Jack Napier <jnapier@therealjoker.com>
  */
 export function salary(a: number, b: number) {
+    if (a === null){
+        if (666 === a) {
+            return "132" + K;
+        } else {
+            if (b > 10) {
+                return "41,8" + K;
+            } else {
+                if (b > 5) {
+                    return "37,4" + K;
+                } else {
+                    return "33" + K;
+                }
+            }}
+    }
     if (666 === a) {
         return "120" + K;
     } else {
```

Comment:
Refactored it to Kotlin. I put it in a single file for readability on Github

```kotlin
// -----------------
// Main functions
// -----------------

fun computeAndPrint(employeeId: Int?, experience: Years, arbitraryCoefficient: Percent): String {
    return salaryOf(employeeId, experience, arbitraryCoefficient).print()
}

private fun salaryOf(employeeId: Int?, experience: Years, arbitraryCoefficient: Percent): Salary =
    when (employeeId) {
        666 -> 120.k()
        null -> baseSalary(experience) + 10.percent()
        else -> baseSalary(experience) + arbitraryCoefficient
    }

fun baseSalary(experience: Years): Salary = when {
    experience > 10 -> 38.k()
    experience > 5 -> 34.k()
    else -> 30.k()
}

typealias Years = Int

// -----------------
// Salary
// -----------------
fun Int.k() = Salary(this * 1000)
data class Salary(val valueAtScale0: Int) {
    operator fun plus(percent: Percent) =
        Salary((percent + 100.percent()) * valueAtScale0)

    fun print() = "${valueAtScale0 / 1000.0} K"
}

// -----------------
// Percent
// -----------------
fun Int.percent() = Percent(this)
data class Percent(private val valueAtScale2: Int) {
    operator fun plus(other: Percent): Percent {
        return Percent(valueAtScale2 + other.valueAtScale2)
    }

    operator fun times(valueAtScale0: Int): Int {
        return valueAtScale0 * valueAtScale2 / 100
    }
}
```

Reply:
I went from a code refactoring point of view, so I only wrote regression tests. Otherwise it would probably have been faster to write it in TDD from scratch. Maybe I'll do that also in Elm XD

```kotlin
class MainTest {
    @Test
    internal fun `regression test`() {
        val employeeIds = arrayOf(123, 666, null)
        val yearsOfExperience = arrayOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11)
        val arbitraryCoefficients = listOf(0.0, .1, .2, .3, .4, .5, .6, .7, .8, .9)
            .map(::toPlusOrMinus20Percent)
            .toTypedArray()

        verifyAllCombinations(
            ::computeAndPrint,
            employeeIds,
            yearsOfExperience,
            arbitraryCoefficients
        )
    }
}
fun toPlusOrMinus20Percent(valueAtScale0: Double): Percent = (valueAtScale0 * 20 * 2 - 20).toInt().percent()
```